File: 程序员面试宝典/3.3.堆盘子.py
class StackOfPlates:
    def __init__(self, cap: int):
        self.cap = cap
        self.stacks = []

    def push(self, val: int) -> None:
        if self.cap == 0:
            return
        if not self.stacks or len(self.stacks[-1]) == self.cap:
            self.stacks.append([])

        self.stacks[-1].append(val)

    def pop(self) -> int:
        if not self.stacks:
            return -1

        res = self.stacks[-1].pop(-1)
        if len(self.stacks[-1]) == 0:
            del self.stacks[-1]
        return res

    def popAt(self, index: int) -> int:
        if len(self.stacks) <= index:
            return -1

        res = self.stacks[index].pop(-1)
        if len(self.stacks[index]) == 0:
            del self.stacks[index]
        return res


# Your StackOfPlates object will be instantiated and called as such:
# obj = StackOfPlates(cap)
# obj.push(val)
# param_2 = obj.pop()
# param_3 = obj.popAt(index)

File: 程序员面试金典/3.3.堆盘子.cpp
#include <stack>
#include <vector>
using namespace std;

class StackOfPlates {
public:
    vector<stack<int>> sts;
    int cap;

    StackOfPlates(int cap)
    {
        this->cap = cap;
    }

    void push(int val)
    {
        if (cap == 0) {
            return;
        }

        if (sts.empty() || sts.back().size() == cap) {
            sts.emplace_back(stack<int>());
        }
        sts.back().push(val);
    }

    int pop()
    {
        if (cap == 0 || sts.empty()) {
            return -1;
        }
        int res = sts.back().top();
        sts.back().pop();
        if (sts.back().empty()) {
            sts.pop_back();
        }
        return res;
    }

    int popAt(int index)
    {
        if (cap == 0 || index >= sts.size()) {
            return -1;
        }
        int res = sts[index].top();
        sts[index].pop();
        if (sts[index].empty()) {
            sts.erase(sts.begin() + index);
        }
        return res;
    }
};
